src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const result = [];
+
+  for (const action of actions) {
+    const st = { ...copyState };
+
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          st[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete st[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in st) {
+          delete st[key];
+        }
+        break;
+    }
+
+    copyState = st;
```

Comment:
Reassigning `copyState` to `st` at the end of each loop iteration is unnecessary since `copyState` is not used elsewhere in the loop. Consider removing this line.